trajopt_sco/src/osqp_interface.cpp

```diff
@@ -0,0 +1,270 @@
+#include <constants.h>
+#include <cmath>
+#include <Eigen/SparseCore>
+#include <fstream>
+#include <signal.h>
+
+#include <trajopt_sco/osqp_interface.hpp>
+#include <trajopt_sco/solver_utils.hpp>
+#include <trajopt_utils/logging.hpp>
+#include <trajopt_utils/stl_to_string.hpp>
+
+namespace sco
+{
+double OSQP_INFINITY = std::numeric_limits<double>::infinity();
+
+ModelPtr createOSQPModel()
+{
+  ModelPtr out(new OSQPModel());
+  return out;
+}
+
+OSQPModel::OSQPModel()
+{
+  // Define Solver settings as default
+  osqp_set_default_settings(&osqp_settings_);
+  osqp_settings_.eps_abs = 1e-4;
+  osqp_settings_.eps_rel = 1e-6;
+  osqp_settings_.max_iter = 8192;
+  osqp_settings_.polish = 1;
+
+  // Initialize data
+  osqp_data_.A = nullptr;
+  osqp_data_.P = nullptr;
+  osqp_workspace_ = nullptr;
+}
+
+OSQPModel::~OSQPModel()
+{
+  // Cleanup
+  if (osqp_workspace_ != nullptr)
+    osqp_cleanup(osqp_workspace_);
+  if (osqp_data_.A != nullptr)
+    c_free(osqp_data_.A);
+  if (osqp_data_.P != nullptr)
+    c_free(osqp_data_.P);
+}
+
+Var OSQPModel::addVar(const std::string& name)
+{
+  vars_.push_back(new VarRep(vars_.size(), name, this));
+  lbs_.push_back(-OSQP_INFINITY);
+  ubs_.push_back(OSQP_INFINITY);
+  return vars_.back();
+}
+
+Cnt OSQPModel::addEqCnt(const AffExpr& expr, const std::string& /*name*/)
+{
+  cnts_.push_back(new CntRep(cnts_.size(), this));
+  cnt_exprs_.push_back(expr);
+  cnt_types_.push_back(EQ);
+  return cnts_.back();
+}
+
+Cnt OSQPModel::addIneqCnt(const AffExpr& expr, const std::string& /*name*/)
+{
+  cnts_.push_back(new CntRep(cnts_.size(), this));
+  cnt_exprs_.push_back(expr);
+  cnt_types_.push_back(INEQ);
+  return cnts_.back();
+}
+
+Cnt OSQPModel::addIneqCnt(const QuadExpr&, const std::string& /*name*/)
```

Comment:
I substituted the assert and return with a throw, so that this happens clearly whenever used. At the moment quadratic inequality constraints are never used, and at the moment, among the supported solvers, only `Gurobi` supports them. If necessary, we might add support for QCQP  in the future. I can't come up at the moment with a more elegant solution at the moment.. ideas?